### DISCUSSION TITLE
Wait for next/dynamic chunks to load before displaying the page during client-side navigation

### DISCUSSION THREAD

**Discussion by vajajak** [vercel/next.js#19142] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

The idea behind next/dynamic for component-level code splitting is ingenious. In my case, I'm using it to split different article blocks that I'm pulling from my cms. In case of big packages like react-player, it's useless to load them in article detail pages, where video block is not even present.

Using dynamic components however causes content layout shift (especially on slower connections) in client-side transitions. Unlike the main page bundle, the page is not waiting for chunks to fully download before displaying. In case of large articles, it looks and feels really ugly when content's jumping around as article blocks are loading.

## Describe the solution you'd like

I would like to block displaying of the page until all chunks are fully loaded. This way, no cls happens and page transitions are smooth without content jumping around.
There could be a similar option as is fallback: 'blocking' in incremental static generation.

## Describe alternatives you've considered

I am aware that there is a 'loading' option which lets me create my own loader, but that just doesn't feel sufficient, because:
- In case I wanted to use some facebook-like placeholder, I don't know the block's height in advance
- It feels so un-Next-like to see loading in place of text on pages that are statically pre-generated

## Additional context

For now, this is the reason why I'm staying away from dynamic components. Please let me know if there is already some solution to this. I was, however, unable to find any. Thanks!

<img width="610" alt="Screenshot 2020-11-13 at 12 10 40" src="https://user-images.githubusercontent.com/52152407/99066605-840bd200-25a9-11eb-8249-d09d4151255d.png">

**Comment by bastienrobert** [comment]
Maybe related with #2476 about blocking the router at any moment (and therefore during a dynamic component loading). Did you find any workaround?
Also, it could be cool to add a prefetching option in the router to load the chunks as well.

**Comment by simoncorompt** [comment]
We're encountering the same problem here : being able to create pages from a CMS with a lot of different blocks / layouts, and only have the necessary code from the blocks that are used in our pages.

As @vajajak mentioned, skeletons are not an option mainly because in components coming from a CMS, we often can't guess the size of the content in advance. It creates a lot of visual shifting in the page and we'd prefer to have the page appear in one go.

Being able to specify the "dependencies" of a page as "blocking" might be a pretty good solution imo. Maybe directly in the options passed to next/dynamic? So we could have finer control over which components may or may not be delayed.

This is a major hurdle for any big editorial/e-commerce site with lots of content and flexible layouts managed by a marketing team. With all the modern CMS' pushing towards a flexible model based on slices and re-usable components, I think this is an issue that's really important to address :)

Happy to help if more insight is needed on this!

**Reply by timfuhrmann** [reply]
We're currently facing the same issue/have faced this issue for a while now. May I ask what approach you chose to handle this problem for now?

**Comment by devknoll** [comment]
The issue is that we don't know which dynamic chunks the next page will use before we render it. Theoretically, we could throw an error when we discovered an unloaded chunk to abort the render and try again in the future, but this may not be safe for all Next.js apps or 3rd party component libraries with e.g. render phase side effects that are expecting that render is normally only called once on the client.

Instead of aborting the render, theoretically we could render to a different, hidden container, and only make it visible when everything is loaded. But lots of 1st and 3rd party code assumes that when e.g. `useEffect` or `componentDidMount` are called the component is on the screen, to e.g. trigger some other behavior like a scroll or modal dialog. Such assumptions would be violated here.

Thankfully, all of these problems are actually being addressed by React itself: https://github.com/facebook/react/issues/13206 so this sort of thing should be possible in the near future 😄

**Reply by simoncorompt** [reply]
Thanks for chiming in! You're right on the fact that both the ideas you mention have their limits, so definitely looking forward to the release of concurrent mode and suspense as well! :)

In the meantime I think we'll give a try at option 2 on our side to try to minimize the effects of dynamic imports on the user experience.

**Comment by ayayalar** [comment]
I am experiencing the same problem with dynamic imports. I've tried using the experimental features of React 18. Suspend, concurrent or blocking mode didn't seem to help.
